server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java

```diff
@@ -0,0 +1,167 @@
+
+import reactor.core.publisher.Mono;
+
+public class CachedBlobStore implements BlobStore {
+
+    private static final String DEFAULT_BUCKET = "cassandraDefault";
+
+    private final BlobStoreCache cache;
+    private final BlobStore backend;
+    private final Integer sizeThresholdInBytes;
+    private final BucketName defaultBucket;
+
+    @Inject
+    public CachedBlobStore(BlobStoreCache cache, BlobStore backend,
+                           CassandraCacheConfiguration cacheConfiguration,
+                           @Named(DEFAULT_BUCKET) BucketName defaultBucket) {
+        this.cache = cache;
+        this.backend = backend;
+        this.sizeThresholdInBytes = cacheConfiguration.getSizeThresholdInBytes();
+        this.defaultBucket = defaultBucket;
+    }
+
+    @Override
+    public InputStream read(BucketName bucketName, BlobId blobId) throws ObjectStoreIOException, ObjectNotFoundException {
+        Preconditions.checkNotNull(bucketName, "bucketName should not be null");
+
+        return Mono.just(bucketName)
+            .filter(defaultBucket::equals)
+            .flatMap(ignored ->
+                Mono.from(cache.read(blobId))
+                    .<InputStream>flatMap(bytes -> Mono.fromCallable(() -> new ByteArrayInputStream(bytes))))
+            .switchIfEmpty(Mono.fromCallable(() -> backend.read(bucketName, blobId)))
+            .blockOptional()
+            .orElseThrow(() -> new ObjectNotFoundException(String.format("Could not retrieve blob metadata for %s", blobId)));
+    }
+
+    @Override
+    public Mono<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
+        return Mono.just(bucketName)
+            .filter(defaultBucket::equals)
+            .flatMap(ignored -> Mono.from(cache.read(blobId)))
+            .switchIfEmpty(Mono.from(backend.readBytes(bucketName, blobId)));
+    }
+
+    @Override
+    public Mono<BlobId> save(BucketName bucketName, byte[] bytes, StoragePolicy storagePolicy) {
+        return Mono.from(backend.save(bucketName, bytes, storagePolicy))
+            .flatMap(blobId -> {
+                if (isAbleToCache(bucketName, bytes, storagePolicy)) {
+                    return Mono.from(cache.cache(blobId, bytes)).thenReturn(blobId);
+                }
+                return Mono.just(blobId);
+            });
+    }
+
+    @Override
+    public Publisher<BlobId> save(BucketName bucketName, InputStream inputStream, StoragePolicy storagePolicy) {
+        Preconditions.checkNotNull(inputStream, "InputStream must not be null");
+
+        if (isAbleToCache(bucketName, storagePolicy)) {
+            return saveInCache(bucketName, inputStream, storagePolicy);
+        }
+
+        return backend.save(bucketName, inputStream, storagePolicy);
+    }
+
+    private Publisher<BlobId> saveInCache(BucketName bucketName, InputStream inputStream, StoragePolicy storagePolicy) {
+        PushbackInputStream pushbackInputStream = new PushbackInputStream(inputStream, sizeThresholdInBytes + 1);
```

Comment:
Why +1, can it be a named field ?

Reply:
Can we adopt the same style than here: https://github.com/linagora/james-project/pull/3295#discussion_r412761395

Reply:
Yes, i doing it